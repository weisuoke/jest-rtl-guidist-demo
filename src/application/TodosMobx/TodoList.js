import React from "react";
import PropTypes from "prop-types";
import { observer } from "mobx-react";
import "./TodosMobx.less";

const TodoList = ({ store }) => {
  const {
    todoList,
    inputTodo,
    changeInputTodo,
    addTodo,
    toggleTodo,
    deleteTodo,
  } = store;
  return (
    <div>
      TodoList
      <div>========</div>
      <div>
        <input
          type="text"
          value={inputTodo}
          onChange={(e) => {
            e.persist();
            changeInputTodo(e.target.value);
          }}
        />
        <button
          onClick={() => {
            addTodo(inputTodo);
            changeInputTodo("");
          }}
        >
          添加
        </button>
      </div>
      {todoList.map((todo) => {
        return (
          <div key={todo.id} className="TodosMobxItem">
            <input
              type="checkbox"
              checked={todo.done}
              onClick={() => {
                toggleTodo(todo.id);
              }}
            />
            <div>{todo.content}</div>
            <div
              className="delete"
              onClick={() => {
                deleteTodo(todo.id);
              }}
            >
              x
            </div>
          </div>
        );
      })}
    </div>
  );
};

TodoList.propTypes = {
  store: PropTypes.shape({
    addTodo: PropTypes.func,
    inputTodo: PropTypes.string,
    changeInputTodo: PropTypes.func,
    toggleTodo: PropTypes.func,
    deleteTodo: PropTypes.func,
    todoList: PropTypes.arrayOf(),
  }).isRequired,
};

export default observer(TodoList);
