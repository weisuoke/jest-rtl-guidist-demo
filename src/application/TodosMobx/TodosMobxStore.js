import { observable, action, makeObservable } from "mobx";
import { v4 as uuidv4 } from "uuid";

class TodosMobxStore {
  constructor() {
    makeObservable(this);
  }

  @observable todoList = [
    {
      id: uuidv4(),
      content: "Todo1",
      done: false,
    },
    {
      id: uuidv4(),
      content: "Todo2",
      done: true,
    },
    {
      id: uuidv4(),
      content: "Todo3",
      done: false,
    },
  ];

  @observable inputTodo = "";

  @action changeInputTodo = (value) => {
    this.inputTodo = value;
  };

  @action addTodo = (content) => {
    if (content) {
      this.todoList.push({
        id: uuidv4(),
        content,
        done: false,
      });
    }
  };

  @action toggleTodo = (id) => {
    this.todoList.forEach((todo) => {
      if (todo.id === id) {
        todo.done = !todo.done;
      }
    });
  };

  @action deleteTodo = (id) => {
    this.todoList = this.todoList.filter((todo) => todo.id !== id);
  };
}

const TodosMobxStoreInstance = new TodosMobxStore();

export default TodosMobxStore;

export { TodosMobxStoreInstance };
