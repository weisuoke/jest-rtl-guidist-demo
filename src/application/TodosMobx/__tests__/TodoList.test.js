import React from 'react'
import { shallow } from 'enzyme'
import TodosMobxStore from "../TodosMobxStore";
import { TodoList } from "../TodoList";
import toJson from "enzyme-to-json";

describe("Mobx TodoList", () => {
  // eslint-disable-next-line no-unused-vars
  let storeInstance = ""

  beforeEach(() => {
    storeInstance = new TodosMobxStore()
  })

  it("snapshot", () => {
    storeInstance.todoList.forEach((todo, index) => {
      todo.id = index
    })
    const component = shallow(<TodoList store={storeInstance}/>)

    expect(toJson(component)).toMatchSnapshot()
  })

  it("add Todo", () => {
    const component = shallow(<TodoList store={storeInstance}/>)
    const todoInput = component.find('[data-test-id="todo-input"]')
    const addButton = component.find('[data-test-id="add-button"]')
    todoInput.simulate("change", {
      target: {
        value: "123"
      },
      persist: jest.fn()
    })
    expect(storeInstance.inputTodo).toBe("123")
    addButton.simulate("click")
    expect(storeInstance.inputTodo).toBe("")
    // TODO 这里期望是 4. 但是一直是 3.
    // expect(storeInstance.todoList.length).toBe(4)
  })

})