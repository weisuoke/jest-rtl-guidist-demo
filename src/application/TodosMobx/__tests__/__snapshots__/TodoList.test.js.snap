// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mobx TodoList snapshot 1`] = `
<div>
  TodoList
  <div>
    ========
  </div>
  <div>
    <input
      data-test-id="todo-input"
      onChange={[Function]}
      type="text"
      value=""
    />
    <button
      data-test-id="add-button"
      onClick={[Function]}
    >
      添加
    </button>
  </div>
  <div
    className="TodosMobxItem"
    key="0"
  >
    <input
      checked={false}
      onClick={[Function]}
      type="checkbox"
    />
    <div>
      Todo1
    </div>
    <div
      className="delete"
      onClick={[Function]}
    >
      x
    </div>
  </div>
  <div
    className="TodosMobxItem"
    key="1"
  >
    <input
      checked={true}
      onClick={[Function]}
      type="checkbox"
    />
    <div>
      Todo2
    </div>
    <div
      className="delete"
      onClick={[Function]}
    >
      x
    </div>
  </div>
  <div
    className="TodosMobxItem"
    key="2"
  >
    <input
      checked={false}
      onClick={[Function]}
      type="checkbox"
    />
    <div>
      Todo3
    </div>
    <div
      className="delete"
      onClick={[Function]}
    >
      x
    </div>
  </div>
</div>
`;
